.NET SDK - Phase 2 - Workflows - Initial Comment Period

cretz opened the pull request:
# .NET SDK - Phase 2 - Workflows - Initial Comment Period

This is the proposal for the second phase of the .NET SDK.

* See it [rendered](https://github.com/cretz/temporal-proposals/blob/dotnet-workflows/dotnet/sdk-phase-2.md) (or can use the "rich diff" link in upper right; rendered link will disappear on PR merge)
* Use this PR for comments/discussion. Don't be afraid of making too many comments; if it gets too noisy, I'll just create another PR.
* Any points of contention will be discussed in-person by the SDK team and a decision will be made

bergundy commented on dotnet/sdk-phase-2.md:
```diff
@@ -0,0 +1,484 @@
+    present.
+* Workflow queries are public methods with the non-inheritable `[WorkflowQuery]` attribute
+  * Can be in base class, but if overridden, must also have the attribute
+  * Method cannot return a `Task`, but can accept arguments
```
How hard would it be to add support for `Task` returning queries? I'm asking since we're talking about allowing starting local activities from update validators and queries.

Sushisource replied:
I agree with Chad. I don't think we should even call such things "local activities" if we allow it. We should just say "you can disable the sandbox and do stuff from queries if you really want, just don't mutate state, pinky swear". I don't think making it a "real" local activity really buys people anything, but does wildly complicate internals.

bergundy replied:
I think the reason why we do want to leverage activities here is for dependency injection.
I don't know if we'll want to call the API local activity, I agree that is confusing.
I do see a reason to eventually support returning tasks and also using a mechanism supported by the worker so we can avoid triggering the deadlock detector or in generally blocking the workflow thread.
We don't need to design this now though.
I was just pointing out that this might need to change and wanted to understand how hard it would be to change when needed.

bergundy commented on dotnet/sdk-phase-2.md:
```diff
@@ -0,0 +1,484 @@
+
+    public int MaxConcurrentWorkflowTasks { get; set; } = 100;
+
+    public double NonStickyToStickyPollRatio { get; set; } = 0.2;
```
I'd avoid exposing this for now since we might remove this option in the future from Core.

cretz replied:
Fixed

bergundy commented on dotnet/sdk-phase-2.md:
```diff
@@ -0,0 +1,484 @@
+
+TODO(cretz): For discussion:
+
+* Intentionally leaving off `MaxConcurrentWorkflowTaskPolls`, right?
```
I would for now, let's add it later if we see there's demand and we decide not to do poller autoscaling.

cretz replied:
Fixed

bergundy commented on dotnet/sdk-phase-2.md:
```diff
@@ -0,0 +1,484 @@
+    public static Task<TResult> ExecuteActivityAsync<T, TResult>(
+        Func<T, Task<TResult>> activity, T arg, ActivityStartOptions options) { /*...*/ }
+
+    public static Task ExecuteActivityAsync(
+        string activity, object[] args, ActivityStartOptions options) { /*...*/ }
+
+    public static Task<TResult> ExecuteActivityAsync<TResult>(
+        string activity, object[] args, ActivityStartOptions options) { /*...*/ }
```
I like how you avoided the ambiguity and only do multiple args with stringly typed activities.

cretz replied:
:+1: This was also done client side for workflows, signals, and queries. This was also done in Python (kinda, in a slightly different way).

bergundy commented on dotnet/sdk-phase-2.md:
```diff
@@ -0,0 +1,484 @@
+
+    public static void UpdateSearchAttributes(params SearchAttributeUpdate[] updates) { /*...*/ }
+
+    public static IDictionary<string, Payload> RawMemo { get; }
```
Can we have a memo getter that doesn't return raw payloads, something that will embed the configured data converter so users don't have to decode payloads themselves?

cretz replied:
Yes, from the notes after this code:

```
* We may add helpers for getting individual memo values given a type to deserialize to
  * We don't expose a data converter to the user in a workflow at this time, so it is wise to make these helpers
```

cretz replied:
Added explicit API for this

bergundy commented on dotnet/sdk-phase-2.md:
```diff
@@ -0,0 +1,484 @@
+
+* `ChildWorkflowHandle` has `GetResultAsync()` to match `WorkflowHandle` in client, but this could just be a `Task`
+  property. Should it be? Other .NET APIs tend to do this.
+* In client I call options `WorkflowStartOptions`, not only for `StartWorkflowAsync` but also for
+  `ExecuteWorkflowAsync`. I have a similar pattern here with activities and children. Should I continue this or should I
+  remove the `Start` word?
```
Both Java and TS call this `WorkflowOptions`, I don't see why we need to mention `Start`, there's nothing to disambiguate (ATM).

cretz replied:
Ok, I may make this change everywhere then.

bergundy commented on dotnet/sdk-phase-2.md:
```diff
@@ -0,0 +1,484 @@
+* In client I call options `WorkflowStartOptions`, not only for `StartWorkflowAsync` but also for
+  `ExecuteWorkflowAsync`. I have a similar pattern here with activities and children. Should I continue this or should I
+  remove the `Start` word?
+* Should I rename `Now` to `UtcNow` to match .NET since it's non-local? Or does matching `DateTime.UtcNow` not matter
```
I'd vote for matching .NET here and trading off verbosity for clarity.

cretz replied:
It's less about verbosity and more about that "now" is just a clearer term across our SDKs. But I can change to `UtcNow` which may be a tad less clear/discoverable but that should be ok.

cretz replied:
Fixed

bergundy commented on dotnet/sdk-phase-2.md:
```diff
@@ -0,0 +1,484 @@
+    `ReplayAsync` regardless of class name.
+* `WorkflowReplayerOptions` shares many options with `TemporalWorkerOptions`, but not enough (and a few extra from
+  client) to justify any code sharing
+* `ReplayWorkflowAsync` is thread safe and we'll document this instead of making an overload of this that accepts an
+  iterator
+* Unlike Python we choose not to provide the history on the `WorkflowReplayResult` because we don't aggregate for the
+  user (they have LINQ ways to run many in parallel and they can correlate as desired)
```
As long as the experience without accepting an iterator is nice, I'm all of less APIs.

bergundy reviewed (approved):
LGTM, didn't have any major comments.

Sushisource commented on dotnet/sdk-phase-2.md:
```diff
@@ -0,0 +1,484 @@
+* Every workflow must have the non-inheritable `[Workflow]` attribute
+  * Optional `Name` positional property can be given. Default is the unqualified type name with the "I" prefix removed
+    if it's an interface and the next letter is also capital
+    * So what does `IDWorkflow` become? Yes, `DWorkflow` :-( Weighed as an acceptably rare cost for obvious benefit
```
I still maintain just having to type the name is better than the inevitable set of users that will run into this and be surprised.

cretz replied:
It's pretty normal to use the derived name by default (in .NET and in our SDKs). I don't think this is ever gonna matter in practice. This one gotcha is only for those that are intentionally violating .NET standards (something I've found is very rare for iface names).

Sushisource commented on dotnet/sdk-phase-2.md:
```diff
@@ -0,0 +1,484 @@
+      * Problem is instance activities will want to use `Refs` too.
+      * If I put `Refs` in `Temporalio.Workflows` then at least I can know that all uses are for workflows
+        * But then it'd clash with a `Temporalio.Activities.Refs` or do we want something like either
+          `Refs.ForWorkflow<T>` or `WorkflowRefs.Create<T>`?
```
I think `WorklowRefs.Create<T>` is best

cretz replied:
Better than `Refs.Create<T>()`?

cretz replied:
Was decided to separate this into `WorkflowRefs` and `ActivityRefs` that will also validate the workflow annotations.

cretz replied:
Fixed

Sushisource commented on dotnet/sdk-phase-2.md:
```diff
@@ -0,0 +1,484 @@
+
+    public void SetSignalHandler<T>(string name, Func<T, Task> handler) { /*...*/ }
+
+    public void RemoveSignalHandler(string name) { /*...*/ }
```
What happens with this and tryget (as well as for queries) if the handler was defined by `[WorkflowSignal]` ?

cretz replied:
The handler map is built from the attributes but technically is mutable through these methods. So yes, you can technically remove a handler added via an attribute. And of course getting a handler by name works no matter how that handler was added. In that sense attribute-based handlers (our preference) are just sugar on this. It doesn't make much sense for a user to remove an attribute-added handler IMO, but at least it's a simple to understand concept. This is the same way it works in Python atm too. I hope nobody ever calls dynamic handler things FWIW.

cretz commented:
Added updates based on design session. Specifically:

* Fixed typos
* For clarity use an instance-method for an activity
* Change `Refs` to `ActivityRefs` and `WorkflowRefs`
* Confirm `NonStickyToStickyPollRatio` and `MaxConcurrentWorkflowTaskPolls` will not be user-visible for now
* Add typed memo getters
* Conform to new typed search attribute proposal
* Changed `Workflow.Now` to `Workflow.UtcNow`
* Updated `ContinueAsNewException` to show it has to be created with a static function for type safety reasons

Will merge early next week if no more feedback comes

cretz commented on dotnet/sdk-phase-2.md:
````diff
@@ -0,0 +1,484 @@
+}
+```
+
+Some things to not about the above code (much copied from phase 1):
````
Fixed
